### DISCUSSION TITLE
How to combine repeating combinators with recognize

### DISCUSSION THREAD

**Discussion by Enduriel** [winnow-rs/winnow#555] [Q&A] [answered]
Rust is requiring me to annotate types for the `separated` combinator before I call `.recognize()`, but I have no idea what types to use.

What I'm trying to do is match eg a version string with a variable number of digits, so `1.0.0` but also `1.0.0.0`, my initial instinct is to do something like:

```rs
fn parse_version<'s>(input: &mut &'s str) -> PResult<TextBlock<'s>> {
    let version = separated(2..=3, digit1, '.')
        .recognize()
        .parse_next(input)?;

    Ok(TextBlock::Version(version))
}
```

but the problem is that I get
```
type annotations needed
cannot satisfy `_: winnow::stream::Accumulate<&'s str>`
```
and I'm not sure how to satisfy system here as the types are quite complicated. I couldn't find anything in examples, issues or other discussions about a similar issue. Obviously I could just manually match 2/3 repetitions without using separated but I'm hoping to understand how to solve this in a way I could apply to other situations.

If there's a satisfactory answer to this I thing it would be really useful to use as an example for the `recognize` doc.

**Comment by epage** [comment, marked as answer]
We `recognize()` on a `Accumulate` type in several places in the docs but likely not in a place to discover it.  Between the calls, we add `.map(|()| ())`.

**Reply by epage** [reply]
Posted #556

**Reply by Enduriel** [reply]
Thanks, that looks right and ty for adding it to the doc.
